File: programmers/키패드 누르기.py
def solution(numbers, hand):
    answer = ''
    left = [1, 4, 7]
    right = [3, 6, 9]
    middle = [2, 5, 8, 11]
    r_hand = 10
    l_hand = 12
    for num in numbers:
        if num == 0:
            num = 11
        if num in left: # 왼손 리스트에 속한 경우
            l_hand = num
            answer += 'L'
        elif num in right: # 오른손 리스트에 속한 경우
            r_hand = num
            answer += 'R'
        else:
            # 거리 체크
            dist1 = abs(num - l_hand) // 3 + abs(num - l_hand) % 3
            dist2 = abs(num - r_hand) // 3 + abs(num - r_hand) % 3
            if dist1 > dist2:
                answer += 'R'
                r_hand = num
            elif dist1 < dist2:
                answer += 'L'
                l_hand = num
            else: # 거리가 같을 경우
                if hand == 'right':
                    answer += 'R'
                    r_hand = num
                else:
                    answer += 'L'
                    l_hand = num
    return answer

File: programmers_level1/키패드_누르기.cpp
#include <string>
#include <vector>
#include <cmath>
#define ASTERISK 10
#define HASH 12

using namespace std;

int distance(int from, int to) {
    int dist{ 0 };
    dist += abs((from-1) / 3 - (to-1) / 3); // 열
    dist += abs((from-1) % 3 - (to-1) % 3); // 행
    return dist;
}

string solution(vector<int> numbers, string hand) {
    string answer = "";
    int left_hand{ ASTERISK }, right_hand{ HASH };
    bool used_left{ false };

    for (int target : numbers) {
        if (target == 1 || target == 4 || target == 7) {
            used_left = true;
        } else if (target == 3 || target == 6 || target == 9) {
            used_left = false;
        } else {
            if (target == 0) target = 11;
            int dist_left{ distance(left_hand, target) }, dist_right{ distance(right_hand, target) };
            if (dist_left < dist_right) {
                used_left = true;
            } else if (dist_left > dist_right) {
                used_left = false;
            } else { // same distance
                if (hand == "left") {
                    used_left = true;
                } else {
                    used_left = false;
                }
            }
        }
        if (used_left) {
            left_hand = target;
            answer += 'L';
        } else {
            right_hand = target;
            answer += 'R';
        }
    }

    return answer;
}